[SNAP-2582] allow for server-auth-provider to be NONE

sumwale opened the pull request:
## Changes proposed in this pull request

- fixing issues with setting server-auth-provider to NONE while having auth-provider=LDAP
  - remove dependence of Misc.isSecurityEnabled and similar methods on peer authentication service
    rather use the regular auth service; also cleaned up reading/handling of auth service properties
  - internal connections (including GfxdConnectionWrapper) should use the properties from the
    regular authentication service rather than the peer auth service
  - update handling of boot properties in AuthenticationServiceBase so that the relevant security
    properties are extracted for both regular and peer auth services (so that changes listed above
        can use the regular auth service and obtain the security properties)
  - corrected other issues like setup of default schema in GfxdConnectionWrapper
- corrected auth error in shutdown by using the boot properties from the auth service for shutdown
- updated the default LDAP user search filter to match against many more LDAP implementations
- switched security tests to randomly set server-auth-provider to NONE to test for above changes
- added full wait for stats layer before firing any DDL/system procedure to fix occasional
  failures in stats tests

## Patch testing

precheckin, new tests and manual testing

## Is precheckin with -Pstore clean?

store side is now clean; there is a failure due to lock timeout on snappydata side which happens without these changes and will be tracked separately

## ReleaseNotes changes

Document the possibility of setting "-server-auth-provider=NONE" to disable peer-to-peer authentication

## Other PRs 

https://github.com/SnappyDataInc/snappydata/pull/1167

Commit:
[SNAP-2582] allow for server-auth-provider to be NONE

- fixing issues with setting server-auth-provider to NONE while having auth-provider=LDAP
  - remove dependence of Misc.isSecurityEnabled and similar methods on peer authentication service
    rather use the regular auth service; also cleaned up reading/handling of auth service properties
  - internal connections (including GfxdConnectionWrapper) should use the properties from the
    regular authentication service rather than the peer auth service
  - update handling of boot properties in AuthenticationServiceBase so that the relevant security
    properties are extracted for both regular and peer auth services (so that changes listed above
        can use the regular auth service and obtain the security properties)
  - corrected other issues like setup of default schema in GfxdConnectionWrapper
- corrected auth error in shutdown by using the boot properties from the auth service for shutdown
- updated the default LDAP user search filter to match against many more LDAP implementations
- switched security tests to randomly set server-auth-provider to NONE to test for above changes
- added full wait for stats layer before firing any DDL/system procedure to fix occaional
  failures in stats tests

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/Misc.java:
```diff
@@ -169,6 +170,18 @@ public final static GemFireStore getMemStoreBootingNoThrow() {
+    final GemFireStatSampler sampler = getDistributedSystem().getStatSampler();
+    if (sampler != null) {
+      try {
+        sampler.waitForInitialization();
```
Would you want to use the timed wait version instead? and fail if for whatever reason that does not get initialized.

sumwale replied:
Ok, will make it timeout after ackWaitThreshold (which is 15s by default after which normal messages spit out "15 seconds elapsed...").

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -174,9 +174,9 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 
   /** DOCUMENT ME! */
-  private AuthenticationService authenticationService;
+  private AuthenticationServiceBase authenticationService;
```
In all the change in this file you have just used the abstract "AuthenticationServiceBase" instead of the interface "AuthenticationService". Any specific reason?

sumwale replied:
Callers now use the getBootCredentials which is not in the interface.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/distributed/message/StatementExecutorMessage.java:
```diff
@@ -146,7 +146,10 @@ public StatementExecutorMessage(ResultCollector<Object, T> collector,
         getTimeStatsSettings(lcc), abortOnLowMemory);
-    if (source != null && !Property.DEFAULT_USER_NAME.equals(defaultSchema)) {
+    // skip setting defaultSchema when it is for APP system user
+    if (source != null && !(Property.DEFAULT_USER_NAME.equals(defaultSchema) &&
```
Do we see a situation where in a secured snappy cluster can have even default schema being used? I mean an authenticated user's default schema will be his username space only. Skipping it is fine but I don't think this will actually be the case. Right? If yes then probably we can just set the default schema blindly.

sumwale replied:
This is for an unsecured cluster where the user is APP and then the defaultSchema can be skipped because that is what remote connection will use.

Earlier remote side (GfxdConnectionWrapper) would always use peer authentication service properties for the connection but that was a problem and now it uses the dbOwner instead (which credentials can be different if peer authentication service is different from normal one).

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/impl/jdbc/EmbedStatement.java:
```diff
@@ -2064,6 +2064,10 @@ boolean executeStatement(Activation a,
           connForRemote = false;
           tran = null;
         }
+        if (act != null) {
```
Yeah again. Why is this required and should we use the timed wait version instead?

sumwale replied:
Sampler initialization is in background and I saw failures in stats tests where it will never get the expected region mbeans created because the sampler was not initialized till then. So this will ensure that region mbeans are successfully created for all tables. Specifically saw that occasionally the "InetAddress.getCanonicalHostName" would get stuck for a while because it depends on remote DNS server responses. I have changed that too to use normal toString() which includes both host name and IP address to make it sufficiently unique for stats display purpose.

kneeraj reviewed (approved):
LGTM.

Minor clarification and suggestion.

Commit:
review comments and more fixes + improvements

- use max ackWaitThreshold to wait for sampler initialization
- mask generic passwords from DDL string logging (patterns: password... or pwd...)
- correct default schema send for APP in DistributedProcedureCallFunction
- submit a background process for management beans for embedded connection during
  bootup itself (rather than skipping completely) to fix couple of failures in management tests
